src/compilation_database.cpp

```diff
@@ -0,0 +1,292 @@
+#include "compilation_database.hpp"
+
+#include "Context.hpp"
+#include "File.hpp"
+#include "Logging.hpp"
+
+#include <third_party/llvm_yaml_escape.hpp>
+
+#include <string.h>
+
+#include <vector>
+
+using Logging::log;
+
+namespace {
+
+bool
+needs_to_be_escaped(std::string_view str)
+{
+  for (auto i = str.begin(), e = str.end(); i != e; ++i) {
+    uint8_t v = static_cast<uint8_t>(*i);
+    if (v < 0x20 || v > 0x7E || *i == '\\' || *i == '"') {
+      return true;
+    }
+  }
+  return false;
+}
+
+void
+append_quoted_and_escaped(std::string& out, std::string_view sv)
+{
+  out.append("\"");
+  if (needs_to_be_escaped(sv)) {
+    out.append(llvm::yaml::escape(sv));
+  } else {
+    out.append(sv.data(), sv.length());
+  }
+  out.append("\"");
+}
+
+enum Action : int {
+  unknown = 0,
+  skip,
+  skip_arg_also,
+  keep,
+  keep_and_split_at,
+};
+
+struct ArgStatus
+{
+  static ArgStatus
+  unknown()
+  {
+    return {Action::unknown, 0};
+  }
+
+  static ArgStatus
+  keep()
+  {
+    return {Action::keep, 0};
+  }
+
+  static ArgStatus
+  skip()
+  {
+    return {Action::skip, 0};
+  }
+
+  static ArgStatus
+  skip_arg_also()
+  {
+    return {Action::skip_arg_also, 0};
+  }
+
+  static ArgStatus
+  keep_and_split_at(size_t pos)
+  {
+    return {Action::keep_and_split_at, int(pos)};
+  }
+
+  Action action;
+  int split_pos;
+};
+
+// Code which generates the compilation database in clang:
+// lib/Driver/ToolChains/Clang.cpp  Clang::DumpCompilationDatabase
+
+// Clang removes "-x" and "-M" arguments from the output and splits various
+// arguments, e.g. "-I/path" is always converted to ["-I", "/path"].
+//
+// See include/clang/Driver/CC1Options.td and include/clang/Driver/Options.td
+// for the group definitions, and see the "JoinedOrSeparate" for the splitting:
+//
+// relevant are: CC1Option Preprocessor_Group clang_i_Group I_Group M_Group
+//
+// cat include/clang/Driver/Options.td |
+//   grep -E 'JoinedOrSeparate.*(CC1Option|(Preprocessor|M|clang_i|I)_Group)'
+//
+// This function duplicates this behavior.
```

Comment:
Do we really need to duplicate Clang's behavior? If ccache is called to wrap a certain compilation command , wouldn't it be OK to just put that compilation command as-is in the compilation database?

I would like to avoid having compiler-specific knowledge in ccache unless it's strictly necessary. More complexity means more bugs and more code to update and maintain. Specifically, it would be good not to have to make updates in this area to match future Clang versions.